### ISSUE TITLE
Keyword argument warning on Ruby 2.7 for #and_call_original

### ISSUE THREAD

**Issue by jaynetics** [rspec/rspec-mocks#1306]
### Subject of the issue
`#and_call_original` with keyword arguments causes the following warning:

> ~/.rbenv/versions/2.7.0/lib/ruby/gems/2.7.0/gems/rspec-mocks-3.9.0/lib/rspec/mocks/message_expectation.rb:101: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call

### Your environment
* Ruby version: 2.7.0
* rspec-expectations version: 3.9.0

### Steps to reproduce
```ruby
require 'rspec'

class Foo
  def initialize(bar: nil); end
end

RSpec.describe Foo do
  it 'works' do
    expect(Foo).to receive(:new).with(bar: :qux).and_call_original
    Foo.new(bar: :qux)
  end
end

RSpec::ExampleGroups::Foo.run
```

### Expected behavior
No warning is emitted.

### Actual behavior
A warning is emitted.

**Comment by JonRowe** [comment]
@jaynetics sorry it took so long but #1324 should take of this warning

**Comment by jonspalmer** [comment]
Any news on this one?

**Comment by pirj** [comment]
@jonspalmer Thanks for the reminder!

**Comment by G-Rath** [comment]
Any further news on this? We'd like to get on Ruby 2.7.1 and this is the only "blocker" (not a showstopper, but would be nice to have resolved).

**Comment by bestwebua** [comment]
The same issue, any updates?

**Comment by JonRowe** [comment]
Its on the list of things the team are working on and as always PRs are welcome.

**Comment by ioquatix** [comment]
It looks like this is now broken on Ruby 3. I'm happy to take a look, but I guess I don't know what the current status is.

**Comment by pirj** [comment]
@ioquatix No work is done on this front yet, we're focusing on preparing 4.0 release. Would be happy to include kwargs delegation fixes along with it.

If you can spare some time, here are notable things to get you going. Most importantly, it seems that using `ruby2_keywords` is the only choice. See:
 - https://github.com/rspec/rspec-expectations/issues/1241#issuecomment-749710208
 - https://github.com/rspec/rspec-expectations/issues/1221#issuecomment-716133588

**Comment by JonRowe** [comment]
Work has been done on this but it slipped through the cracks, in part because we don't have specs for a lot of the keyword argument usage. Any keyword argument specs that can be provided which now fail I'll happily fix and take priority over releasing RSpec 4 so people can use Ruby 3 while we work on it.

**Comment by pirj** [comment]
@ioquatix Have you had a chance to take a look? No stress, we're going to pair on this with @benoittgt, just to avoid doing duplicate work.

**Comment by JonRowe** [comment]
Applying this demonstrates the issue:
```diff
--- a/spec/rspec/mocks/and_call_original_spec.rb
+++ b/spec/rspec/mocks/and_call_original_spec.rb
@@ -12,6 +12,10 @@ RSpec.describe "and_call_original" do
           yield x, :additional_yielded_arg
         end

+        if RSpec::Support::RubyFeatures.kw_args_supported?
+          class_eval("def meth_3(a: ''); a; end")
+        end
+
         def self.new_instance
           new
         end
@@ -59,6 +63,12 @@ RSpec.describe "and_call_original" do
       expect(value).to eq([:submitted_arg, :additional_yielded_arg])
     end

+    it 'supports keyword arguments' do
+      expect(instance).to receive(:meth_3).and_call_original
+      value = instance.meth_3(a: :a)
+      expect(value).to eq(:a)
+    end
+
```

**Comment by JonRowe** [comment]
(Or at least it generates the warning, I can't run rspec-mocks on Ruby 3.0.0 locally at the moment due to a `executable-hooks` 🤷  issue)

**Comment by ioquatix** [comment]
I have not had time to look into it yet so thank you for your efforts.

**Comment by swiknaba** [comment]
double `.as_null_object` seems to also suffer from this problem under Ruby 3.0 --haven't had time to dig deeper yet though. Example spec: https://github.com/WeTransfer/zip_tricks/blob/v5.5.0/spec/zip_tricks/streamer_spec.rb#L526 and here the CI failure: https://github.com/WeTransfer/zip_tricks/pull/103/checks?check_run_id=1649814176#step:7:434

**Comment by JonRowe** [comment]
This specific issue is fixed by #1385.

@swiknaba can you open a new issue for anything with `.as_null_object` ideally with an isolated reproduction (e.g. a snippet thats just RSpec and plain Ruby) if you have the time, I don't have the time to go read your build / project source right now.

**Comment by swiknaba** [comment]
> This specific issue is fixed by #1385.
> 
> @swiknaba can you open a new issue for anything with `.as_null_object` ideally with an isolated reproduction (e.g. a snippet thats just RSpec and plain Ruby) if you have the time, I don't have the time to go read your build / project source right now.

Done! https://github.com/rspec/rspec-mocks/issues/1396  --for the history: the bad guy is `receive`, not `as_null_object`. sorry.
